### PR TITLE
feat(stack): add mergify stack squash command

### PR DESCRIPTION
**Pull request by jd**
Add "mergify stack squash SRC... into TARGET [-m MESSAGE]" which reorders
SRC commits adjacent to TARGET and folds them in without opening an
editor. Without -m, TARGET's message is kept (via the fixup rebase
action). With -m, the rebase inserts an `exec git commit --amend -m ...`
line right after the last fixup source, which runs while HEAD still
points at the combined target — so prepare-commit-msg re-attaches the
Change-Id normally.

Errors when SRC == TARGET, when a SRC is duplicated, or when a SRC or
TARGET prefix is unknown. Supports --dry-run.

Click does not allow two positionals when one is variadic, so the CLI
takes a single tokens tuple and parses "SRC... into TARGET" manually
(_parse_squash_tokens); parser edge cases are covered by unit tests.

Also documents the squash command in skills/mergify-stack/SKILL.md.

### CODE REVIEW

**Comment by jd** [comment]
This pull request is part of a [Mergify stack](https://docs.mergify.com/stacks/):

| # | Pull Request | Link | |
|--:|---|---|---|
| 1 | feat(stack): add mergify stack squash command | [#1266](https://github.com/Mergifyio/mergify-cli/pull/1266) | 👈 |

File: skills/mergify-stack/SKILL.md [hunk 1]
```diff
@@ -38,6 +39,7 @@ A branch is a stack. Keep stacks short and focused:
 | `gh pr edit --title "..."` | Edit the commit message, then `mergify stack push` | PR title/body are overwritten from commit messages on every push |
 | `gh pr merge` or `gh pr close` | PR lifecycle is fully managed — do nothing | PR lifecycle is fully managed by the stack tool |
 | `git commit` on `main` | `mergify stack new <name>` first | `mergify stack push` will fail on the default branch |
+| `git rebase -i` to squash/fixup commits | `mergify stack fixup <SHA>` or `mergify stack squash A B into X [-m "..."]` | Non-interactive — works inside LLM/agent sessions; no editor spawned |
```
Comment (jd) [inline skills/mergify-stack/SKILL.md]: fixup is not part of this PR

File: skills/mergify-stack/SKILL.md [hunk 2]
```diff
@@ -55,6 +57,10 @@ mergify stack move X first   # Move commit X to the top of the stack
 mergify stack move X last    # Move commit X to the bottom of the stack
 mergify stack move X before Y  # Move commit X before commit Y
 mergify stack move X after Y   # Move commit X after commit Y
+mergify stack fixup X              # Fold commit X into its parent (drops X's message)
+mergify stack fixup X Y Z          # Fold each into its parent (multi-fixup)
```
Comment (jd) [inline skills/mergify-stack/SKILL.md:63]: ditto

**Comment by jd** [comment]
### Revision history

| # | Type | Changes | Date |
|---|------|---------|------|
| 1 | initial | `6e5abfd` | 2026-04-20 14:08 UTC |
| 2 | content | [`6e5abfd → 96320ff`](https://github.com/Mergifyio/mergify-cli/compare/6e5abfd531fadc7a8df4bc1e4d8d71a3323500b5...96320ff63987ef595b8c2f72afa1294a00ecd5fe) | 2026-04-20 14:08 UTC |
| 3 | rebase | [`96320ff → 24e8198`](https://github.com/Mergifyio/mergify-cli/compare/96320ff63987ef595b8c2f72afa1294a00ecd5fe...24e8198a646228127793608612043d2025be1981) | 2026-04-21 08:01 UTC |
| 4 | content | [`24e8198 → dd97d10`](https://github.com/Mergifyio/mergify-cli/compare/24e8198a646228127793608612043d2025be1981...dd97d1084ea078e0df7d41a1038c8d933505aad5) | 2026-04-21 08:53 UTC |

File: skills/mergify-stack/SKILL.md (hunk 1 above)
Comment (jd) [reply skills/mergify-stack/SKILL.md]: Fixed. Moved the fixup-related SKILL.md additions (Core Convention, Commands line, Common Mistakes row, stash-reminder entry) into #1265 where the fixup command is introduced. This PR's SKILL.md diff now only documents squash.

File: skills/mergify-stack/SKILL.md (hunk 2 above)
Comment (jd) [reply skills/mergify-stack/SKILL.md:63]: Fixed — same change as above: fixup content moved to #1265, squash-only content kept here.

File: mergify_cli/stack/squash.py [hunk 3]
```diff
@@ -85,5 +86,72 @@ async def stack_squash(
...
+
+    if dry_run:
+        console.print("Dry run — no changes made", style="green")
+        return
+
+    exec_after_sha: str | None = None
+    exec_command: str | None = None
+    if message is not None:
+        exec_after_sha = src_shas[-1]
+        exec_command = f"git commit --amend -m {shlex.quote(message)}"
```
[the remark this answers is not in the document]
Comment (jd) [reply mergify_cli/stack/squash.py]: Good catch. Switched to passing the message via a temp file + `git commit --amend -F <file>`, so multi-line messages (or any shell-sensitive characters) can't break the rebase-todo. Added `test_squash_with_multiline_message` covering a subject + multi-line body with embedded quotes as regression coverage.

**Review by JulianMaurin — APPROVED** [review]

**Review by kozlek — APPROVED** [review]
